infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java

```diff
@@ -0,0 +1,216 @@
+
+    /**
+     * Returns a new BatchProcessor for a stream log.
+     *
+     * @param streamLog      stream log for writes (can be in memory or file)
+     * @param sealEpoch All operations stamped with epoch less than the epochWaterMark are
+     *                       discarded.
+     * @param streamLog stream log for writes (can be in memory or file)
+     * @param sync    If true, the batch writer will sync writes to secondary storage
+     */
+    public BatchProcessor(StreamLog streamLog, long sealEpoch, boolean sync) {
+        this.sealEpoch = sealEpoch;
+        this.sync = sync;
+        this.streamLog = streamLog;
+        operationsQueue = new LinkedBlockingQueue<>();
+        processorService.submit(this::processor);
+    }
+
+    /**
+     * Add a task to the processor.
+     * @param type The request type
+     * @param msg The request message
+     * @return returns a future result for the request, if it expects one
+     */
+    public <T> CompletableFuture <T> addTask(@Nonnull Type type, @Nonnull CorfuPayloadMsg msg) {
+        BatchWriterOperation<T> operation = new BatchWriterOperation<>(type, msg);
+        operationsQueue.add(operation);
+        return operation.getFutureResult();
+    }
+
+    private void processor() {
+
+        if (!sync) {
+            log.warn("batchWriteProcessor: writes configured to not sync with secondary storage");
+        }
+
+        try {
+            BatchWriterOperation lastOp = null;
+            int processed = 0;
+            List<BatchWriterOperation> res = new LinkedList<>();
+
+            while (true) {
+                BatchWriterOperation currOp;
+
+                if (lastOp == null) {
+                    currOp = operationsQueue.take();
+                } else {
+                    currOp = operationsQueue.poll();
+
+                    if (currOp == null || processed == BATCH_SIZE
+                            || currOp == BatchWriterOperation.SHUTDOWN) {
+                        streamLog.sync(sync);
+                        log.trace("Sync'd {} writes", processed);
+
+                        for (BatchWriterOperation operation : res) {
+                            if (!operation.getFutureResult().isCompletedExceptionally()
+                            && !operation.getFutureResult().isCancelled()) {
+                                // At this point we need to complete the requests
+                                // that completed successfully (i.e. haven't failed)
+                                operation.getFutureResult().complete(operation.getRetVal());
+                            }
+                        }
+                        res.clear();
+                        processed = 0;
+                    }
+                }
+
+                if (currOp == null) {
+                    lastOp = null;
+                } else if (currOp == BatchWriterOperation.SHUTDOWN) {
+                    log.trace("Shutting down the write processor");
```

Comment:
Should this be warning level? so it can hint on why operations are not being processed in case of any issues.

Reply:
Done.